Fix/duplicate identifier validation for address

fabiodmota opened the pull request:
<!-- 
 'fix: Duplicate Identifier Validation on New Business Partners'
-->

## Description
Added service validation when a request is made with identifier with the same values at the address level.
Added unit test for testing the error return


https://github.com/eclipse-tractusx/bpdm/issues/413

## Pre-review checks

Please ensure to do as many of the following checks as possible, before asking for committer review:

- [x] DEPENDENCIES are up-to-date. [Dash license tool](https://github.com/eclipse/dash-licenses). Committers can open IP issues for restricted libs.
- [x] [Copyright and license header](https://eclipse-tractusx.github.io/docs/release/trg-7/trg-7-02) are present on all affected files

martinfkaeser commented:
I found a case that is not handled yet.

The expected behaviour from the issue is:

> When POSTing new legal addresses which have the same identifiers (value and type are the same) to the Pool, the Pool should reject these business partners due to duplicate identifiers.

But now this fix only prohibits identifiers that are absolutely identical (regarding all 3 fields: type, value, issuingBody). 
If there are two requests where identifier type and value are equal, but issuingBody is different, this is not detected and both entities are accepted.

I have modified your LE test which should now detect this case too in branch `fix/Duplicate_Identifier_Validation_MKA` which you can cherry-pick from.

martinfkaeser commented on bpdm-pool-api/src/main/kotlin/org/eclipse/tractusx/bpdm/pool/api/model/response/PoolErrorCodes.kt:
```diff
@@ -66,7 +66,8 @@ enum class AddressCreateError : ErrorCode {
     RegionNotFound,
-    IdentifierNotFound
+    IdentifierNotFound,
+    MainAddressDuplicateIdentifier,
```
Shouldn't it be just called `AddressDuplicateIdentifier`?

martinfkaeser commented on bpdm-pool/src/main/kotlin/org/eclipse/tractusx/bpdm/pool/service/RequestValidationService.kt:
```diff
@@ -225,6 +257,29 @@ class RequestValidationService(
+    ): Set<LegalEntityIdentifierDto> {
+        val allIdentifiers = legalEntityRequests.flatMap { it.identifiers }
+
+        return allIdentifiers.groupBy { it }
```
This should only consider the fields `type` and `value`. Maybe we could use data class `IdentifierCandidateKey` for that?

martinfkaeser commented on bpdm-pool/src/main/kotlin/org/eclipse/tractusx/bpdm/pool/service/RequestValidationService.kt:
```diff
@@ -78,13 +76,43 @@ class RequestValidationService(
+    fun validateLegalEntityDuplicates(legalEntity: LegalEntityDto, duplicateIdentifiers: Set<LegalEntityIdentifierDto>, entityKey: String?) : Collection<ErrorInfo<LegalEntityCreateError>> {
+        val errorList = mutableListOf<ErrorInfo<LegalEntityCreateError>>()
+        duplicateIdentifiers.forEach { duplicate ->
+            if (legalEntity.identifiers.contains(duplicate)) {
```
Same as `findDuplicateLegalEntityIdentifiers` this should only consider the fields type and value.

martinfkaeser commented:
I think `validateLegalEntityUpdates()` and `validateAddressUpdates()` should also check for duplicate identifiers in the request, right?

martinfkaeser commented:
I think the address identifiers nested as legal address or site main address need to be checked as well.

fabiodmota commented:
i added a new comment to all new developments @martinfkaeser thank you for the feedback

martinfkaeser commented:
I combined the duplicate checks for identifiers in the DB and in the request.
Then I added a check & test for the case of a duplicate legal address identifier that is already in the DB.
Also added findDuplicates() extension function to CollectionUtils.

fabiodmota commented:
@martinfkaeser looks good to me , for me its resolved also

nicoprow reviewed (changes requested):
Just saw in the commit history, the second commit has a missing main commit description. Please add a header description to this commit to put the the additional description in context.

fabiodmota commented:
> Just saw in the commit history, the second commit has a missing main commit description. Please add a header description to this commit to put the the additional description in context.

Fixed , and refactored commit header @nicoprow
